### DISCUSSION TITLE
How to invoke/run workflows from client application or api call using workflowdefinitionId

### DISCUSSION THREAD

**Discussion by NiveditaGalwade** [elsa-workflows/elsa-core#4908] [Q&A]
Consider we have created multiple custom workflows using programmatic way(implemented by IWorkflow), we wanted to call/invoke workflow using workflowdefinitionId. 
Currently we are invoking workflow using  BuildAndStartWorkflowAsync method of IBuildsAndStartsWorkflow, like 
RunWorkflowResult ?  workFlowResult = await _workflowBuilder.BuildAndStartWorkflowAsync<CustomWorkFlow>(input: workFlowInputItems);

We wanted to invoke using workflowdefinitionId, to make it generic.
Can someone please suggest way to achieve this?

Also we wanted to maintain the workflow state and resume particular activity from where it was suspended?

Thank you in advance!!!

**Comment by glime-ai** [comment]
To invoke a workflow using the `workflowDefinitionId`, you can use the `IWorkflowRuntime` service. This service allows you to start workflows by providing the definition ID and any input you want to pass to the workflow. Here's an example of how you can do this:

```csharp
var workflowRuntime = serviceProvider.GetRequiredService<IWorkflowRuntime>();
var input = new Dictionary<string, object>
{
    // Add your input parameters here
};
var startWorkflowOptions = new StartWorkflowRuntimeOptions(input: input);
var result = await workflowRuntime.StartWorkflowAsync(workflowDefinitionId, startWorkflowOptions);
```

To maintain the workflow state and resume a particular activity from where it was suspended, you need to use bookmarks. Bookmarks are a way to pause the workflow execution and resume it later. When an activity creates a bookmark, the workflow instance will be in a suspended state until the bookmark is resumed.

To resume a workflow from a specific bookmark, you can use the `IWorkflowLaunchpad` service. Here's an example of how you can resume a workflow using a bookmark:

```csharp
var workflowLaunchpad = serviceProvider.GetRequiredService<IWorkflowLaunchpad>();
var bookmark = new Bookmark
{
    // Set the bookmark properties here
};
var result = await workflowLaunchpad.CollectAndExecuteWorkflowsAsync(
    new WorkflowsQuery(
        nameof(YourActivity), 
        bookmark
    ),
    new WorkflowInput(input),
    cancellationToken
);
```

Please note that the exact code and services might vary depending on the version of Elsa you are using. The provided examples are based on the information available in the context sources.

For more detailed information and examples, you can refer to the Elsa documentation and guides provided in the context sources.

**Reply by NiveditaGalwade** [reply]
Getting this exception
System.Exception: 'Specified workflow definition and version does not exist'
even though i am providing 
var startWorkflowOptions = new StartWorkflowRuntimeOptions() { Input = input , VersionOptions = Elsa.Common.Models.VersionOptions.Published, CancellationTokens = CancellationToken.None  };
and workFlowId = 'CustomWorkflowId'

**Reply by NiveditaGalwade** [reply]
I am also not getting getting workflowdefinitions by using below method

 private readonly IWorkflowDefinitionStore _workflowDefinitionStore;
 var result =  await _workflowDefinitionStore.FindByDefinitionIdAsync(definitionId, VersionOptions.Published, cancellationToken);
 return result;
 
result is empty, its not null.

**Reply by robverhoef-united** [reply]
Make sure you have added the workflow with AddWorkflow<#workflowclassname#>().

**Reply by NiveditaGalwade** [reply]
I have already registered workflow using AddWorkflow<#workflowclassname#>

**Comment by NiveditaGalwade** [comment]
This issue is resolved
